drivers/SmartThings/zigbee-button/src/zigbee-multi-button/robb/init.lua

```diff
@@ -0,0 +1,183 @@
+local device_management = require "st.zigbee.device_management"
+local battery_defaults = require "st.zigbee.defaults.battery_defaults"
+
+local zcl_clusters = require "st.zigbee.zcl.clusters"
+local Basic = zcl_clusters.Basic
+local Level = zcl_clusters.Level
+local OnOff = zcl_clusters.OnOff
+local PowerConfiguration = zcl_clusters.PowerConfiguration
+local capabilities = require "st.capabilities"
+
+--[[
+The ROBB Wireless Remote Control has 4 or 8 buttons. They are arranged in two columns:
+
+All buttons on the left side support 'pressed' (OnOff > ON) and 'up_hold' (Level > MoveStepMode.UP).
+All buttons on the right side support 'pressed' (OnOff > OFF) and 'down_hold' (Level > MoveStepMode.DOWN).
+
+Each button-row represents one endpoint. The 8x remote control has four endpoints, the 4x remote control has two.
+That means each endpoint has two buttons.
+--]]
+
+local SWITCH_GROUP_CONFIGURE = "is_group_configured"
+local SWITCH8_NUM_ENDPOINT = 4
+local SWITCH8_NUM_BUTTONS = 8
+local SWITCH4_NUM_ENDPOINT = 2
+local SWITCH4_NUM_BUTTONS = 4
+
+local WIRELESS_REMOTE_FINGERPRINTS = {
+  { mfr = "ROBB smarrt", model = "ROB_200-007-0" },
+  { mfr = "ROBB smarrt", model = "ROB_200-008-0" }
+}
+
+local function can_handle(opts, driver, device, ...)
+  for _, fingerprint in ipairs(WIRELESS_REMOTE_FINGERPRINTS) do
+    if device:get_manufacturer() == fingerprint.mfr and device:get_model() == fingerprint.model then
+      return true
+    end
+  end
+  return false
+end
+
+-- Map left column buttons to endpoints
+local EP_BUTTON_ON_COMPONENT_MAP = {
+  [0x01] = "button1",
+  [0x02] = "button3",
+  [0x03] = "button5",
+  [0x04] = "button7"
+}
+
+-- Map right column buttons to endpoints
+local EP_BUTTON_OFF_COMPONENT_MAP = {
+  [0x01] = "button2",
+  [0x02] = "button4",
+  [0x03] = "button6",
+  [0x04] = "button8"
+}
+
+local build_button_handler = function(driver, device, zb_rx)
+    local button_name
+    local additional_fields = { state_change = true }
+
+    -- Fetch correct button name
+    if zb_rx.body.zcl_header.cmd.value == 0x01 then
```

Comment:
So, if you'll notice in the `button_utils` function you've based this off of, it's a function _factory_. I expect what you want to do is also have a factory, but pass in the correct component map to be used.

But, if you wanted to leverage the existing systems a bit more, I might suggest setting the `component_to_endpoint` function up like this:

```
function endpoint_to_component(device, ep)
local EP_MAP = {
  [0x01] = "button1",
  [0x02] = "button3",
  [0x03] = "button5",
  [0x04] = "button7",
  [0xF1] = "button2",
  [0xF2] = "button4",
  [0xF3] = "button6",
  [0xF4] = "button8"
}
return device.profile.components[EP_MAP[ep]]
end
```
and then, make `build_button_handler` actually a factory that takes one flag `addF` that is a boolean used to determine whether you add `0xF0` to the value of `zb_rx.address_header.src_endpoint.value` before calling `device:emit_event_for_endpoint()` or not.

so your handlers would look like:

```
      [OnOff.ID] = {
        [OnOff.server.commands.Off.ID] = build_button_handler(true),
        [OnOff.server.commands.On.ID] = build_button_handler(false)
      }
```

does that make sense?

Another option is to leave the function largely as is, but eliminate the maps and just use either 

`"switch"..(zb_rx.address_header.src_endpoint.value * 2)`

or

`"switch"..(zb_rx.address_header.src_endpoint.value * 2 - 1)`

depending on which is relevant

Reply:
solved with 714d055